moshi/src/main/java/com/squareup/moshi/ObjectJsonWriter.java

```diff
@@ -131,7 +140,20 @@ public Object root() {
   }
 
   @Override public JsonWriter value(double value) throws IOException {
-    return value(Double.valueOf(value));
+    long longValue = (long) value;
+    if (Double.compare(longValue, value) == 0) {
+      return value(longValue);
```

Comment:
despite having prepared this PR I’m still not 100% convinced that this is the best behavior.

If a caller emits a sequence `[0.0d, 0.5d, 1.0d, 1.5d]` we’ll create JSON of `[0L, 0.5d, 1L, 1.5d]`.

It might be simpler to only do clever stuff in `value(Number)`, and to limit the clever stuff to promoting to a preferred type (Long over Integer, Double over Float). In either case this stuff is tricky.